### PR TITLE
Qutrit Rotations #2 (TRY)

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR builds on previous PR's working on qutrit simulation by adding a parametrized Y rotation operation. This change builds the foundation for making parametrized qutrit circuits, and thus, allows testing variational qutrit algorithms with PennyLane.

**Description of the Change:**
* Added `TRY()` operation, which applies an Y rotation to a subspace specified by the user. The subspace determines which 2 of 3 one-qutrit basis states the operation applies to.

**Benefits:**
* Users can now perform a parametrized Y rotation on qutrit devices.

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -8,6 +8,10 @@
   given subspace.
   [(#2845)](https://github.com/PennyLaneAI/pennylane/pull/2845)
 
+* Added the `TRY` qutrit rotation operator, which allows applying a Pauli X rotation on a
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:11]:
```suggestion
* Added the `TRY` qutrit rotation operator, which allows applying a Pauli Y rotation on a
```

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 2]
```diff
@@ -298,6 +354,13 @@ def test_control_wires(op, control_wires):
     assert op.control_wires == control_wires
 
 
+@pytest.mark.parametrize(
+    "op, obs, grad_fn",
+    [
+        (qml.TRX, qml.GellMann(0, 3), lambda phi: -np.sin(phi)),
+        (qml.TRY, qml.GellMann(0, 8), lambda phi: 0 * phi),
```
Comment (timmysilv) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:361]: `0 * phi`? I'd say what about just `0`, but maybe instead can we update the test to a more meaningful circuit wrt. TRY?

File: pennylane/ops/qutrit/parametric_ops.py [hunk 3]
```diff
@@ -169,3 +169,152 @@ def adjoint(self):
...
+    parameter_frequencies = [(0.5, 1)]
+
+    # Internal dictionary to map subpsaces to Gell-Mann observable for the generator
+    _index_dict = {(0, 1): 2, (0, 2): 5, (1, 2): 7}
+
+    def generator(self):
+        return qml.s_prod(-0.5, qml.GellMann(self.wires, index=self._index_dict[self.subspace]))
+
+    def __init__(
+        self, phi, wires, subspace=[0, 1], do_queue=True, id=None
```
Comment (timmysilv) [inline pennylane/ops/qutrit/parametric_ops.py:240]: same concern as before, can we update the default value for subspace?

File: pennylane/ops/qutrit/parametric_ops.py [hunk 4]
```diff
@@ -169,3 +169,152 @@ def adjoint(self):
...
+
+        Returns:
+            tuple[int]: subspace on which operator acts
+        """
+        return tuple(sorted(self._subspace))
+
+    @staticmethod
+    def compute_matrix(
+        theta, subspace=(0, 1)
+    ):  # pylint: disable=arguments-differ,dangerous-default-value
```
Comment (timmysilv) [inline pennylane/ops/qutrit/parametric_ops.py:264]: can you try removing this pylint warning? and if it's still needed despite tuples being immutable, can we ensure that it's pylint making the mistake and not us?

**Review by timmysilv — COMMENTED** [review]
looks great, just a few small things

**Review by timmysilv — APPROVED** [review]
🥳 🥳 (assuming tests pass)

File: pennylane/ops/qutrit/parametric_ops.py [hunk 5]
```diff
@@ -169,3 +169,148 @@ def adjoint(self):
...
+        self._hyperparameters = {
+            "subspace": self._subspace,
+        }
+        super().__init__(phi, wires=wires, do_queue=do_queue, id=id)
+
+    @property
+    def subspace(self):
+        """The single-qutrit basis states which the operator acts on
+
+        This property returns the 2D subspace on which the operator acts. This subspace
```
Comment (glassnotes) [inline pennylane/ops/qutrit/parametric_ops.py:250]: These first two sentences say effectively the same thing, so can remove one.

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 6]
```diff
@@ -298,6 +354,13 @@ def test_control_wires(op, control_wires):
     assert op.control_wires == control_wires
 
 
+@pytest.mark.parametrize(
+    "op, obs, grad_fn",
+    [
+        (qml.TRX, qml.GellMann(0, 3), lambda phi: -np.sin(phi)),
+        (qml.TRY, qml.GellMann(0, 1), lambda phi: np.cos(phi)),
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:361]: Unrelated to this PR, but I wonder if we should consider making one of the arguments to `GellMann` a mandatory keyword argument. Every time I use it, I forget which is the index, and which is the wire!

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 7]
```diff
@@ -307,152 +370,153 @@ class TestGrad:
...
-        assert np.allclose(jac, -1.0 * jnp.sin(np.diag(phi)), atol=tol, rtol=0)
+        assert np.allclose(jac, np.diag(grad_fn(phi)), atol=tol, rtol=0)
 
     @pytest.mark.torch
     @pytest.mark.parametrize("phi", npp.linspace(0, 2 * np.pi, 7))
     @pytest.mark.parametrize("diff_method", diff_methods)
-    def test_trx_differentiability_torch(self, phi, diff_method, tol):
-        """Test that TRX is differentiable with Torch and the gradient is correct"""
+    def test_differentiability_torch(self, op, obs, grad_fn, phi, diff_method, tol):
+        """Test that parametric operations are differentiable with Torch and the gradient is correct"""
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:446]:
```suggestion
        """Test that parametrized operations are differentiable with Torch and the gradient is correct"""
```

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 8]
```diff
@@ -307,152 +370,153 @@ class TestGrad:
...
 
-        assert qml.math.isclose(grad, -1.0 * np.sin(phi), atol=tol, rtol=0)
+        assert qml.math.isclose(grad, grad_fn(phi), atol=tol, rtol=0)
 
     @pytest.mark.torch
     @pytest.mark.parametrize("diff_method", diff_methods)
-    def test_trx_differentiability_torch_broadcasted(self, diff_method, tol):
-        """Test that differentiation of TRX in Torch with broadcasting works."""
+    def test_differentiability_torch_broadcasted(self, op, obs, grad_fn, diff_method, tol):
+        """Test that differentiation of parametric operations in Torch with broadcasting works."""
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:464]:
```suggestion
        """Test that differentiation of parametrized operations in Torch with broadcasting works."""
```

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 9]
```diff
@@ -307,152 +370,153 @@ class TestGrad:
...
-        assert qml.math.allclose(jac, -1.0 * torch.sin(torch.diag(phi)), atol=tol, rtol=0)
+        assert qml.math.allclose(jac, np.diag(grad_fn(phi)), atol=tol, rtol=0)
 
     @pytest.mark.tf
     @pytest.mark.parametrize("phi", npp.linspace(0, 2 * np.pi, 7))
     @pytest.mark.parametrize("diff_method", diff_methods)
-    def test_trx_differentiability_tf(self, phi, diff_method, tol):
-        """Test that TRX is differentiable with TensorFlow and the gradient is correct"""
+    def test_differentiability_tf(self, op, obs, grad_fn, phi, diff_method, tol):
+        """Test that parametric operations are differentiable with TensorFlow and the gradient is correct"""
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:484]:
```suggestion
        """Test that parametrized operations are differentiable with TensorFlow and the gradient is correct"""
```

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 10]
```diff
@@ -307,152 +370,153 @@ class TestGrad:
...
 
-        assert qml.math.isclose(res, -1.0 * tf.sin(phi), atol=tol, rtol=0)
+        assert qml.math.isclose(res, grad_fn(phi), atol=tol, rtol=0)
 
     @pytest.mark.tf
     @pytest.mark.parametrize("diff_method", diff_methods)
-    def test_trx_differentiability_tf_broadcasted(self, diff_method, tol):
-        """Test that differentiation of TRX in TensorFlow with broadcasting works."""
+    def test_differentiability_tf_broadcasted(self, op, obs, grad_fn, diff_method, tol):
+        """Test that differentiation of parametric operations in TensorFlow with broadcasting works."""
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:505]:
```suggestion
        """Test that differentiation of parametrized operations in TensorFlow with broadcasting works."""
```

**Review by glassnotes — APPROVED** [review]
Looks great, @mudit2812 ! 🎉  Very easy review, now that `TRX` is in.

File: tests/ops/qutrit/test_qutrit_parametric_ops.py (hunk 6 above)
Comment (mudit2812) [reply tests/ops/qutrit/test_qutrit_parametric_ops.py:361]: I would like that as well, but I have two grievances:
* Making `index` a positional argument implies that it is trainable when it isn't supposed to be.
* Keeping it as a keyword argument but placing it before `wires` also breaks convention as keyword arguments are supposed to be placed *after* `wires`.

**Pull request merged**